File: week07/7/[241015](GfG)_QuickSort.py
"""Quick Sort
    1. Divide : Pivot을 기준으로 왼쪽, 오른쪽 2개의 list로 분류한다.
    2. 하나의 list를 sort하는 방법 : low, high index를 활용한다.
        pivot을 제외하고 low는 맨 왼쪽, high는 맨 오른쪽에서 시작함.
        low는 왼->오, high는 오->왼으로 움직이다가
            pivot < list[low] 이면 멈춤.
            pivot > list[high] 이면 멈춤
        low, high가 둘 다 멈춘 시점에는 low, high가 역전되지 않았다면, list[low], list[high]를 swap.
        >> 위 과정은 low >= high가 되는 시점까지 진행한다.(엇갈리는 시점)
    3. 2번과정이 끝나면 Pivot과 high를 바꾼다.
        Recursive하게 호출한 하위 partition에서 새로운 pivot을 설정한다.
"""

"""Quick Sort의 Divide and Conquer
    1. Divide : Pivot을 기준으로 왼쪽은 작은값, 오른쪽은 큰 값들의 partition으로 만든다.
                더이상 divide하지 않는 시점은 partition의 크기가 0 또는 1
    2. Conquer : 들어온 partition을 sort한다. (low, High index를 관리하면서 pivot 기준으로 값 분류하는 과정)
    3. Combine : 1,2가 끝난 2개의 partitions를 결합한다.
"""

class Solution:
    def partition(self,arr,low,high):
        pivotValue = arr[high];
        i = low;
        
        for j in range(low, high):
            if arr[j] <= pivotValue:
                temp = arr[j]
                arr[j] = arr[i]
                arr[i] = temp
                i+=1
        
        temp2 = arr[i]
        arr[i] = pivotValue
        arr[high] = temp2
        
        return i
        
    #Function to sort a list using quick sort algorithm.
    def quickSort(self,arr,low,high):
        if low < high:
            pivot_done = self.partition(arr, low, high)
            self.quickSort(self, low, pivot_done-1);
            self.quickSort(self, pivot_done+1, high)
            
            
    

File: week07/7/[241015](GfG)_QuickSort.cpp
/* <QuickSort>
    1. quickSort : arr[low:high]를 Sort한 후, 분류가 끝난 pivot의 좌-list, 우-list 각각으로 나누어주는 과정. 
        base case : 더이상 divde하지 않는 과정 (길이가 0 || 1)
        recursive :
            left : arr[low:pivot_done]
            right : arr[pivot_done+1 : high]
    2. partition( : arr[low:high] 구간을 pivot을 기준으로 정렬하는 함수
        : 1) last element를 pivotValue로 설정. : arr[high]
        : 2) i, j를 활용한 sort. pivot 좌측은 작은값, pivot 우측은 큰 값.
                i : pivot보다 작은 요소들의 위치
                j : low~high-1 까지를 순회하는 역할
                만약  arr[j] <= pivotValue  :   arr[i], arr[j] swap, 그리고 i++
        : 3) pivotValue 위치를 좌/pivot/우 사이로 옮겨줌.(arr[i+1]와 swap)
        : 4) return high; (pivotValue가 현재 위치한 index)

*/

#include <iostream>
using namespace std;

class Solution{
    public:
    void quickSort(int arr[], int low, int high){
        if (low < high) {
            int pivot_done = partition(arr, low, high);
            quickSort(arr, low, pivot_done-1);
            quickSort(arr,pivot_done+1, high);
        }
    }
    public:
    int partition(int arr[], int low, int high){
        //1) 
        int pivotValue = arr[high];
        int i = low;

        //2)
        for (int j = low; j < high; j++){
            if (arr[j] <= pivotValue){
                int temp = arr[i];
                arr[i] = arr[j];
                arr[j] = temp;

                i++;
            }
        }

        //3
        int temp2 = arr[i];
        arr[i] = pivotValue;
        arr[high] = temp2; // arr[i]값을 기존 pivotValue가 위치했던 자리로.

        //4
        return i;
    }
};